### PR TITLE
fix(compiler-cli): produce smaller source maps for templates

### LINKED ISSUE #19537: Sourcemaps in ngc 5 seems wrong

### PR DESCRIPTION
**Pull request by chuckjaz**
Assocating each template node with a the generated TypeScript
generated overly verbose source maps. Changed to creating a
source map entry per unique source span instead of each
unique template ast node.

Fixes: #19537

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
```

## What is the current behavior?

Issue Number: #19537

The source maps produced are accurate but very verbose.

## What is the new behavior?

The source maps are accurate but produce no longer produce redundant source mappings.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/compiler-cli/src/transformers/node_emitter.ts [hunk 1]
```diff
@@ -92,17 +92,47 @@ class _NodeEmitterVisitor implements StatementVisitor, ExpressionVisitor {
 
   getNodeMap() { return this._nodeMap; }
 
-  private record<T extends ts.Node>(ngNode: Node, tsNode: T|null): RecordedNode<T> {
-    if (tsNode && !this._nodeMap.has(tsNode)) {
-      this._nodeMap.set(tsNode, ngNode);
-      if (!this._mapped.has(ngNode)) {
-        this._mapped.add(ngNode);
+  updateSourceMap(statements: ts.Statement[]) {
```
Comment (tbosch) [inline packages/compiler-cli/src/transformers/node_emitter.ts]: I guess it's not simple, but could you think about some tests that make sure this works as expected?

File: packages/compiler-cli/src/transformers/node_emitter.ts [hunk 2]
```diff
@@ -92,17 +92,47 @@ class _NodeEmitterVisitor implements StatementVisitor, ExpressionVisitor {
...
+    let lastStartNode: ts.Node|undefined = undefined;
+    let lastNode: ts.Node|undefined = undefined;
+    let lastRange: ts.SourceMapRange|undefined = undefined;
+
+    const record = () => {
```
Comment (tbosch) [inline packages/compiler-cli/src/transformers/node_emitter.ts]: Maybe rename: `recordLastSourceMapRange`

File: packages/compiler-cli/src/transformers/node_emitter.ts [hunk 3]
```diff
@@ -92,17 +92,47 @@ class _NodeEmitterVisitor implements StatementVisitor, ExpressionVisitor {
...
+        } else {
+          ts.setSourceMapRange(lastStartNode, lastRange);
+          ts.setEmitFlags(lastStartNode, ts.EmitFlags.NoTrailingSourceMap);
+          ts.setSourceMapRange(lastNode, lastRange);
+          ts.setEmitFlags(lastNode, ts.EmitFlags.NoLeadingSourceMap);
+        }
+      }
+    };
+
+    const map = (tsNode: ts.Node) => {
```
Comment (tbosch) [inline packages/compiler-cli/src/transformers/node_emitter.ts]: maybe rename to `visitNode`? Makes it more obvious that you are just visiting all nodes.

File: packages/compiler-cli/src/transformers/node_emitter.ts [hunk 4]
```diff
@@ -92,17 +92,47 @@ class _NodeEmitterVisitor implements StatementVisitor, ExpressionVisitor {
...
+    let lastStartNode: ts.Node|undefined = undefined;
+    let lastNode: ts.Node|undefined = undefined;
+    let lastRange: ts.SourceMapRange|undefined = undefined;
+
+    const record = () => {
+      if (lastRange && lastStartNode && lastNode) {
+        if (lastStartNode == lastNode) {
+          ts.setSourceMapRange(lastNode, lastRange);
+        } else {
+          ts.setSourceMapRange(lastStartNode, lastRange);
```
Comment (tbosch) [inline packages/compiler-cli/src/transformers/node_emitter.ts]: Maybe add a comment that describes the situation that the `EmitFlags` are preventing.

**Review by tbosch — APPROVED** [review]

File: packages/compiler-cli/src/transformers/node_emitter.ts (hunk 1 above)
Comment (tbosch) [reply packages/compiler-cli/src/transformers/node_emitter.ts]: Maybe just for 2 situations that were are trying to prevent now?

Comment (chuckjaz) [reply packages/compiler-cli/src/transformers/node_emitter.ts]: done

**Review by tbosch — APPROVED** [review]

**Pull request closed without merging**